src/transformation/visitors/switch.ts

```diff
@@ -1,61 +1,139 @@
-    }, undefined as lua.IfStatement | undefined);
+            // Compute the condition for the if statement
+            if (!ts.isDefaultClause(clause)) {
+                if (condition) {
+                    // Coalesce skipped statements
+                    condition = lua.createBinaryExpression(
+                        condition,
+                        lua.createBinaryExpression(
+                            switchVariable,
+                            context.transformExpression(clause.expression),
+                            lua.SyntaxKind.EqualityOperator
+                        ),
+                        lua.SyntaxKind.OrOperator
+                    );
+                } else {
+                    // Next condition
+                    condition = lua.createBinaryExpression(
+                        switchVariable,
+                        context.transformExpression(clause.expression),
+                        lua.SyntaxKind.EqualityOperator
+                    );
+                }
 
-    if (concatenatedIf) {
-        statements.push(concatenatedIf);
-    }
+                // Skip empty clauses
+                if (clause.statements.length === 0) continue;
 
-    const hasDefaultCase = statement.caseBlock.clauses.some(ts.isDefaultClause);
-    statements.push(lua.createGotoStatement(`${switchName}_${hasDefaultCase ? "case_default" : "end"}`));
+                // Declare or assign condition variable
+                statements.push(
+                    isInitialCondition
+                        ? lua.createVariableDeclarationStatement(conditionVariable, condition)
+                        : lua.createAssignmentStatement(
+                              conditionVariable,
+                              lua.createBinaryExpression(conditionVariable, condition, lua.SyntaxKind.OrOperator)
+                          )
+                );
+                isInitialCondition = false;
+            } else {
+                assert(!isInitialCondition, "Default clause should never be the initial condition");
+            }
 
-    for (const [index, clause] of statement.caseBlock.clauses.entries()) {
-        const labelName = `${switchName}_case_${ts.isCaseClause(clause) ? index : "default"}`;
-        statements.push(lua.createLabelStatement(labelName));
-        statements.push(lua.createDoStatement(context.transformStatements(clause.statements)));
-    }
+            // Push if statement for case
+            statements.push(
+                lua.createIfStatement(
+                    conditionVariable,
+                    lua.createBlock(context.transformStatements(clause.statements))
+                )
+            );
+
+            // Clear condition for next clause
+            condition = undefined;
+        }
 
-    statements.push(lua.createLabelStatement(`${switchName}_end`));
+        // Amalgamate the default w/ fallthrough clauses and execute if nothing else executed above
```

Comment:
The code below seems like some black magic, and this comment is not really helping understand it, it might be making it worse actually.

Reply:
Is it the word `amalgamate`...?

I can work to comment through this section better. Do you have any suggestions on how to make it feel less black magic?

Essentially, we have to reconstruct the default clause and combine all of the statements from the clauses that the default can fall into.

Without a goto statement, this was the only approach I could dream up, as discovered by the initial try at this where combining clauses was proven to pass all tests too.

Reply:
> Essentially, we have to reconstruct the default clause and combine all of the statements from the clauses that the default can fall into.

This would be a good start, maybe also explaining what (and why) end is.